httpcache_test.go

```diff
@@ -316,6 +329,59 @@ func TestDontStorePartialRangeInCache(t *testing.T) {
 	}
 }
 
+func TestCacheOnlyIfBodyRead(t *testing.T) {
+	resetTest()
+	{
+		req, err := http.NewRequest("GET", s.server.URL, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		resp, err := s.client.Do(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if resp.Header.Get(XFromCache) != "" {
+			t.Fatal("XFromCache header isn't blank")
+		}
+		// We do not read the body
+		resp.Body.Close()
+	}
+	{
+		req, err := http.NewRequest("GET", s.server.URL, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		resp, err := s.client.Do(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer resp.Body.Close()
+		if resp.Header.Get(XFromCache) != "" {
+			t.Fatalf("XFromCache header isn't blank")
+		}
+	}
+}
+
+func TestOnlyReadBodyOnDemand(t *testing.T) {
+	resetTest()
+
+	req, err := http.NewRequest("GET", s.server.URL+"/infinite", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	resp, err := s.client.Do(req) // This shouldn't hang forever.
+	if err != nil {
+		t.Fatal(err)
+	}
+	buf := make([]byte, 0)
+	_, err = resp.Body.Read(buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+	resp.Body.Close()
+	close(s.done)
```

Comment:
Wouldn't it be a good idea to move `close(s.done)` to `teardown`, just before `s.server.Close()`, so that it mirrors `s.done = make(chan struct{})` being in `setup`?

That way, other tests (in the future) could also rely on the "/infinite" endpoint.